client/src/composables/historyDragDrop.ts

```diff
@@ -0,0 +1,173 @@
+            (!createNew && !destinationHistoryId.value) ||
+            !getDragItems().length ||
+            processingDrop.value
+    );
+
+    function getDragItems() {
+        const storeItems = eventStore.getDragItems();
+        // Filter out any non-history or `DatasetCollectionElement` items
+        return storeItems?.filter(
+            (item) => isHistoryItem(item) || isDatasetElement(item as DCESummary)
+        ) as DraggableHistoryItem[];
+    }
+
+    function onDragEnter(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        dragTarget.value = e.target;
+        showDropZone.value = true;
+    }
+
+    function onDragOver(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        e.preventDefault();
+    }
+
+    function onDragLeave(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        if (dragTarget.value === e.target) {
+            showDropZone.value = false;
+        }
+    }
+
+    async function onDrop() {
+        showDropZone.value = false;
+        if (operationDisabled.value) {
+            return;
+        }
+        processingDrop.value = true;
+
+        let datasetCount = 0;
+        let collectionCount = 0;
+
+        try {
+            const dragItems = getDragItems();
+            const multiple = dragItems.length > 1;
+            const originalHistoryId = fromHistoryId.value as string;
+
+            let historyId;
+            if (destinationHistoryId.value) {
+                historyId = destinationHistoryId.value;
+            } else if (createNew) {
+                await historyStore.createNewHistory();
+                historyId = historyStore.currentHistoryId;
+            } else {
+                historyId = null;
+            }
+            if (!historyId) {
+                throw new Error("Destination history not found or created");
+            }
+
+            // iterate over the data array and copy each item to the current history
+            for (const item of dragItems) {
+                let dataSource: HistoryContentSource;
+                let type: HistoryContentType;
+                let id: string;
+                if (isHistoryItem(item)) {
+                    dataSource = item.history_content_type === "dataset" ? "hda" : "hdca";
+                    type = item.history_content_type;
+                    id = item.id;
+                }
+                // For DCEs, only `DCEDataset`s are droppable, `DCEDatasetCollection`s are not
+                else if (isDatasetElement(item) && item.object) {
+                    dataSource = "hda";
+                    type = "dataset";
+                    id = item.object.id;
+                } else {
+                    throw new Error(`Invalid item type${item.element_type ? `: ${item.element_type}` : ""}`);
+                }
+                await copyDataset(id, historyId, type, dataSource);
```

Comment:
Maybe you'd want to loop over the items to drag first and make sure none of them run into https://github.com/galaxyproject/galaxy/pull/18699/files#r1734725855 ? Otherwise it might be a little messy if a fraction of items is copied and another is not.

Reply:
I have moved the actual copy operation to another loop, so that if errors are caught for any item, we never copy anything from then on.

From what I understand, if there is any error with:
- `historyStore.createNewHistory();`
- `!historyId`
- Invalid item type
- `copyDataset`
- `loadHistoryById`

the error would be handled and Toast-ed in `useHistoryDragDrop.onDrop()`.